js/plugins/mcp/src/client/client.ts

```diff
@@ -0,0 +1,295 @@
+    });
+  }
+
+  /**
+   * Connects to a single MCP server defined by the provided configuration.
+   * If a server with the same name already exists, it will be disconnected first.
+   * Stores the client and transport references internally. Handles connection errors
+   * by marking the server as disabled.
+   * @param serverName The name to assign to this server connection.
+   * @param config The configuration object for the server.
+   */
+  async connect(config: McpServerConfig) {
+    if (this._server) await this._server.transport.close();
+    logger.info(`[MCP Client] Connecting MCP server in client '${this.name}'.`);
+
+    const { transport, type: transportType } = await transportFrom(config);
+    if (!transport) {
+      throw new GenkitError({
+        status: 'INVALID_ARGUMENT',
+        message: `[MCP Client] Could not determine valid transport config from supplied options.`,
+      });
+    }
+
+    let error: string | undefined;
+
+    const client = new Client({ name: this.name, version: this.version });
+    client.registerCapabilities({ roots: {} });
+
+    if (this.isEnabled()) {
+      try {
+        await client.connect(transport);
+      } catch (e) {
+        logger.warn(
+          `[MCP Client] Error connecting server via ${transportType} transport: ${e}`
+        );
+        this.disabled = true;
+        error = (e as Error).toString();
+      }
+    }
+
+    this._server = {
+      client,
+      transport,
+      error,
+    } as McpServerRef;
+  }
+
+  /**
+   * Disconnects the MCP server and removes its registration
+   * from this client instance.
+   */
+  async disconnect() {
+    if (this._server) {
+      logger.info(
+        `[MCP Client] Disconnecting MCP server in client '${this.name}'.`
+      );
+      await this._server.client.close();
+      this._server = undefined;
+    }
+  }
+
+  /**
+   * Temporarily disables a server connection. Closes the underlying transport
+   * but retains the server's configuration. Does nothing if the server is
+   * already disabled.
+   */
+  async disable() {
+    if (!this.isEnabled()) return;
+    if (this._server) {
+      logger.info(`[MCP Client] Disabling MCP server in client '${this.name}'`);
+      this.disabled = true;
+    }
+  }
+
+  /**
+   * Whether this client-server connection is enabled or not.
+   */
+  isEnabled() {
+    return !this.disabled;
+  }
+
+  /**
+   * Re-enables a previously disabled server connection. Does nothing if the
+   * server is not disabled.
+   */
+  async reenable() {
+    if (this.isEnabled()) return;
+    if (this._server) {
+      logger.info(
+        `[MCP Client] Reenabling MCP server in client '${this.name}'`
+      );
+      this.disabled = false;
+    }
+  }
+
+  /**
+   * Closes and then restarts the transport connection for the specified server.
+   * Useful for attempting to recover from connection issues without full
+   * reconfiguration.
+   */
+  async reconnect() {
```

Comment:
Is this the `reconnect` behavior we want:

- `reconnect` does not help establish a connection after calling `disconnect` since the server config is deleted on disconnection. 
- This method documents that it disconnects and reconnects, for recovering connections (and behaves accordingly). 
- If the user wants to "reconnect" after a `disconnect`, they have to call the `connect` method with the serverConfig.

Reply:
Done. (Renamed to `restart`)